### PR TITLE
[Merged by Bors] - chore: further backports for leanprover/lean4#6397

### PR DESCRIPTION
**Pull request by kim-em**

### CODE REVIEW

File: Mathlib/Algebra/FreeMonoid/Count.lean [hunk 1]
```diff
@@ -27,7 +27,8 @@ def countP : FreeAddMonoid α →+ ℕ where
   map_add' := List.countP_append _
 
 theorem countP_of (x : α) : countP p (of x) = if p x = true then 1 else 0 := by
-  simp [countP, List.countP, List.countP.go]
+  change List.countP p [x] = _
```
Comment (jcommelin) [inline Mathlib/Algebra/FreeMonoid/Count.lean:30]: Should we have a simp(?) lemma that brings the LHS in this form?

File: Mathlib/Algebra/FreeMonoid/Count.lean [hunk 2]
```diff
@@ -36,8 +37,8 @@ theorem countP_apply (l : FreeAddMonoid α) : countP p l = List.countP p l := rf
 def count [DecidableEq α] (x : α) : FreeAddMonoid α →+ ℕ := countP (· = x)
 
 theorem count_of [DecidableEq α] (x y : α) : count x (of y) = (Pi.single x 1 : α → ℕ) y := by
-  simp [Pi.single, Function.update, count, countP, List.countP, List.countP.go,
-    Bool.beq_eq_decide_eq]
+  change List.count x [y] = _
```
Comment (jcommelin) [inline Mathlib/Algebra/FreeMonoid/Count.lean:40]: That lemma would then also apply here, I suppose.

File: Mathlib/Algebra/FreeMonoid/Count.lean (hunk 1 above)
Comment (kim-em) [reply Mathlib/Algebra/FreeMonoid/Count.lean:30]: No, I don't think we should, unless someone actually adds the explicit conversion from FreeMonoid to List. As is, it is defeq abuse.

Comment (kim-em) [reply Mathlib/Algebra/FreeMonoid/Count.lean:30]: See https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Mathlib.2FAlgebra.2FFreeMonoid.2FCount.2Elean/near/490048032

Comment (jcommelin) [reply Mathlib/Algebra/FreeMonoid/Count.lean:30]: Aah, thanks for that pointer to zulip.
In that case, I suggest we also leave a comment at the top of the file with the same comments as (or a pointer to) that zulip thread.

**Comment by jcommelin** [comment]
bors d+

**Comment by jcommelin** [comment]
bors merge

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**